icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -744,11 +838,21 @@ MeasureUnit MeasureUnit::withSIPrefix(UMeasureSIPrefix prefix, UErrorCode& statu
 }
 
 int32_t MeasureUnit::getDimensionality(UErrorCode& status) const {
-    return SingleUnitImpl::forMeasureUnit(*this, status).dimensionality;
+    SingleUnitImpl singleUnit = SingleUnitImpl::forMeasureUnit(*this, status);
+    if (U_FAILURE(status)) { return 0; }
+    if (singleUnit.isDimensionless()) {
+        return 0;
+    }
+    return singleUnit.dimensionality;
 }
 
 MeasureUnit MeasureUnit::withDimensionality(int32_t dimensionality, UErrorCode& status) const {
     SingleUnitImpl singleUnit = SingleUnitImpl::forMeasureUnit(*this, status);
+    if (U_FAILURE(status)) { return MeasureUnit(); }
```

Comment:
Optional: you don't actually need lines 851-855 that you added; I would remove them, but you can keep them if you like how it makes getDimensionality more parallel to withDimensionality

Reply:
Done.